### PR TITLE
Regression test for infinite-loop-in-resync

### PR DESCRIPTION
**Pull request by richvdh**
A rather complicated test which demonstrated that the infinite loop was in https://github.com/matrix-org/synapse/issues/13001 was possible.

~~The test currently fails: it's getting stuck waiting for soft-fail, which is something @squahtx is working on.~~

### CODE REVIEW

File: tests/federation_room_join_partial_state_test.go [hunk 1]
```diff
@@ -511,6 +513,192 @@ func TestPartialStateJoin(t *testing.T) {
 			t.Errorf("Did not find derek's m.room.member event in gappy /sync response: %s", err)
 		}
 	})
+
+	// regression test for https://github.com/matrix-org/synapse/issues/13001
+	//
+	// There wass an edge case where, if we initially receive lots of events as outliers,
```
Comment (kegsay) [inline tests/federation_room_join_partial_state_test.go:519]:
```suggestion
	// There was an edge case where, if we initially receive lots of events as outliers,
```

File: tests/federation_room_join_partial_state_test.go [hunk 2]
```diff
@@ -511,6 +513,192 @@ func TestPartialStateJoin(t *testing.T) {
...
+				t.Fatalf("GET /event failed with %d: %s", res.StatusCode, string(eventResBody))
+			}
+			t.Fatalf("timeout waiting for event %s to be received", eventID)
+		}
+
+		// here's the first event which we *ought* to un-partial-state, but won't
+		lateEvent := makeTimelineEvent("late event")
+
+		// next, we want to create 100 outliers. So, charlie creates 100 state events, and
+		// then persuades the SUT to create a backwards extremity using those events as
```
Comment (kegsay) [inline tests/federation_room_join_partial_state_test.go:571]: Unusual acronym 'SUT'?

Comment (richvdh) [reply tests/federation_room_join_partial_state_test.go:571]: "system under test". will reword.

**Comment by richvdh** [comment]
seems like this would benefit from some of the utility functions added in https://github.com/matrix-org/complement/pull/419. I'll update it once that lands.

**Comment by richvdh** [comment]
This is now updated to use the utility functions added in #419. It includes the changes in #428.

File: tests/federation_room_join_partial_state_test.go [hunk 3]
```diff
@@ -563,6 +565,128 @@ func TestPartialStateJoin(t *testing.T) {
 			t.Errorf("Did not find derek's m.room.member event in gappy /sync response: %s", err)
 		}
 	})
+
+	// regression test for https://github.com/matrix-org/synapse/issues/13001
+	//
+	// There was an edge case where, if we initially receive lots of events as outliers,
+	// and they then get de-outliered as partial state events,
+	t.Run("Resync works with many prev_events with partial state", func(t *testing.T) {
```
Comment (squahtx) [inline tests/federation_room_join_partial_state_test.go:573]: Are we missing the other half of the comment?

File: tests/federation_room_join_partial_state_test.go [hunk 4]
```diff
@@ -563,6 +565,128 @@ func TestPartialStateJoin(t *testing.T) {
...
+				}
+				if res.StatusCode == 404 && gjson.GetBytes(eventResBody, "errcode").String() == "M_NOT_FOUND" {
+					t.Logf("Fetching event %s failed with M_NOT_FOUND; will retry", eventID)
+					time.Sleep(1000 * time.Millisecond)
+					continue
+				}
+				t.Fatalf("GET /event failed with %d: %s", res.StatusCode, string(eventResBody))
+			}
+			t.Fatalf("timeout waiting for event %s to be received", eventID)
+		}
```
Comment (squahtx) [inline tests/federation_room_join_partial_state_test.go:602]: It was suggested that we use `WithRetryUntil` here: https://github.com/matrix-org/complement/pull/419#discussion_r928706407 , which looks like identical logic with a different timeout.

We might also want to make this helper available to other tests (and deduplicate the copy from #419)

File: tests/federation_room_join_partial_state_test.go (hunk 3 above)
Comment (squahtx) [reply tests/federation_room_join_partial_state_test.go:573]: Also can we make the name of the test more specific? something along the lines of persisting chains of prev_events out of order maybe

Comment (richvdh) [reply tests/federation_room_join_partial_state_test.go]: I've had a go at fixing the name of the test - perhaps it's better?

File: tests/federation_room_join_partial_state_test.go (hunk 4 above)
Comment (richvdh) [reply tests/federation_room_join_partial_state_test.go]: good idea, thanks. Done.

File: tests/federation_room_join_partial_state_test.go (hunk 3 above)
Comment (squahtx) [reply tests/federation_room_join_partial_state_test.go]: Looks alright!

**Review by squahtx — APPROVED** [review]
lgtm!
